### PR TITLE
Initial implementation of AST

### LINKED ISSUE #19: Literal is stored as a string.
Should consider adding an additional type for the literal including the String, Number or Boolean types as an enum and then encapsulate this on the Token as an Option, instead of defaulting to empty strings when it's not used.

### PR DESCRIPTION
**Pull request by ncatelli**
# Introduction
This PR implements a minimal AST along with representations of each of the expression types. This also includes some refactors of the scanner to cut down on string copies used while storing lexemes in the token by removing it in favor of leveraging the TokenType.

This also introduces a new Literal type for storing an optional Literal with the token.

# Linked Issues
resolves #19 

# Dependencies

# Test
- [x] Tested Locally
- [x] Documented

# Review
- [x] Ready for review
- [x] Ready to merge

# Deployment

### CODE REVIEW

File: librlox/src/ast/expression.rs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+pub struct LiteralExpr {
+    literal: Box<tokens::Token>,
+}
+
+impl LiteralExpr {
+    pub fn new(literal: tokens::Token) -> LiteralExpr {
+        LiteralExpr {
+            literal: Box::new(literal),
```
Comment (ncatelli) [inline librlox/src/ast/expression.rs:71]: No need for this type to be boxed as the size is fixed.

File: librlox/src/ast/tests/expression.rs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+use crate::scanner::tokens::{Literal, Token, TokenType};
+use std::option::Option;
+
+#[test]
+fn test_expression_formatter_should_pretty_print_an_ast() {
+    let expr = Expr::Binary(BinaryExpr::new(
+        Token::new(TokenType::Star, None),
+        Box::new(Expr::Unary(UnaryExpr::new(
+            Token::new(TokenType::Minus, Option::None),
+            Box::new(Expr::Literal(LiteralExpr::new(Token::new(
```
Comment (ncatelli) [inline librlox/src/ast/tests/expression.rs:11]: Will need to remove the box after the above update.

File: librlox/src/ast/tests/expression.rs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+                TokenType::Number,
+                Option::Some(Literal::Number(123.0)),
+            )))),
+        ))),
+        Box::new(Expr::Grouping(GroupingExpr::new(Box::new(Expr::Literal(
```
Comment (ncatelli) [inline librlox/src/ast/tests/expression.rs:16]: Will need to remove the box after the above update.

**Comment by Terkwood** [comment]
Nicely done. This is pretty readable!

**Pull request merged**